src/orquestra/sdk/_base/cli/_dorq/_dumpers.py

```diff
@@ -0,0 +1,89 @@
+################################################################################
+# © Copyright 2023 Zapata Computing Inc.
+################################################################################
+
+"""
+Code that stores values on disk as a result of a CLI command.
+"""
+import json
+import typing as t
+from dataclasses import dataclass
+from functools import singledispatch
+from pathlib import Path
+
+import dill
+
+from orquestra.sdk._base import serde
+from orquestra.sdk.schema import responses
+from orquestra.sdk.schema.ir import ArtifactFormat
+from orquestra.sdk.schema.workflow_run import WorkflowRunId
+
+
+@dataclass(frozen=True)
+class DumpDetails:
+    file_path: Path
+    format: ArtifactFormat
+
+
+@singledispatch
+def _dump_result(result: responses.WorkflowResult, path: Path):
+    raise TypeError(f"Unsupported result type: {type(result)}")
+
+
+@_dump_result.register
+def _(result: responses.JSONResult, path: Path):
+    obj = serde.deserialize(result)
+    with path.open("w") as f:
+        json.dump(obj, f, indent=2)
+
+
+@_dump_result.register
+def _(result: responses.PickleResult, path: Path):
+    obj = serde.deserialize(result)
+    with path.open("wb") as f:
+        dill.dump(obj, f)
+
+
+def _format_extension(format: ArtifactFormat) -> str:
+    if format == ArtifactFormat.JSON:
+        return ".json"
+    elif format == ArtifactFormat.ENCODED_PICKLE:
+        return ".pickle"
+    else:
+        raise ValueError(f"Unsupported format: {format}")
+
+
+class ArtifactDumper:
+    """
+    Writes artifacts to files.
+    """
+
+    def dump(
+        self,
+        value: t.Any,
+        wf_run_id: WorkflowRunId,
+        output_index: int,
+        dir_path: Path,
+    ) -> DumpDetails:
+        """
+        Serialize artifact value and save it as a new file.
+
+        Creates missing directories. Generates filenames based on ``wf_run_id`` and
+        ``output_index``. Figures out the serialization format based on the object. The
+        generated file extension matches the inferred format.
+
+        No standard errors are expected to be raised.
+        """
+        # Dumping to strings instead of directly to byte buffer is suboptimal, but we
+        # wanna use the same serialization procedure as Corq and QE.
+        result_obj = serde.result_from_artifact(
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**question:** ‏‏We do deserialisation (python API) -> serialisation (here) -> deserialisation (`_dump_result`) -> serialisation (`_dump_result`). Is there a better way to handle this?

Reply:
Can you right down the functions calls of how you might do this?

Reply:
I tried many different approaches to reuse as much serde as possible and ended up with f906729125126cef7604f999e0ec34b4.

It doesn't use the same code path as QE but at least we have `dill.dumps()` and `dill.dump()` in the same module.

Reply:
I think it's OK for now but my issue is we're re-serialising still, so if we ever allow a user to pick the serialisation we may overwrite their choice.

Can you add a ticket to tackle this later? I'm thinking that maybe we could add a `WorkflowRun.get_results(raw=True)` which skips deserialisation.